File: sdk/php/tests/LockdownTest.php
<?php

declare(strict_types=1);

namespace Shojiku\Tests;

use PHPUnit\Framework\TestCase;
use Shojiku\Client;
use Shojiku\Configuration;
use Shojiku\Exception\UsageException;
use Shojiku\Lockdown;

/**
 * The lockdown, one clause at a time.
 *
 * A lockdown tested as a whole reports "something was refused" and stops
 * proving which rule did it, so each clause gets its own test.
 */
final class LockdownTest extends TestCase
{
    use EngineFixtures;

    public function testAStrictClientRefusesTheBytesEntrance(): void
    {
        $this->expectException(UsageException::class);
        $this->expectExceptionMessageMatches('/`generateSource` is disabled/');
        $this->client(['strict' => true])->generateSource(template: "version: 0.1.0\nname: x\n");
    }

    public function testAStrictClientRefusesToSignADocumentItDidNotRender(): void
    {
        $strict = $this->client(['strict' => true, 'providers' => ['default' => $this->signer()]]);
        $loaded = $strict->artifact($this->rendered()->bytes());

        $this->expectException(UsageException::class);
        $this->expectExceptionMessageMatches('/only a document rendered from its own template root.*loaded/s');
        $strict->sign($loaded, 'default');
    }

    public function testAStrictClientStillSignsWhatItRenderedItself(): void
    {
        $strict = $this->client(['strict' => true, 'providers' => ['default' => $this->signer()]]);
        $result = $strict->generate('receipt', []);
        $artifact = $result->artifact();
        self::assertNotNull($artifact);

        self::assertTrue($strict->sign($artifact, 'default')->success());
    }

    public function testAStrictClientRefusesAProviderObjectInFavourOfARegisteredName(): void
    {
        $strict = $this->client(['strict' => true, 'providers' => ['default' => $this->signer()]]);

        $this->expectException(UsageException::class);
        $this->expectExceptionMessageMatches('/sign with the name of a provider registered/');
        $strict->sign($this->rendered(), $this->signer());
    }

    public function testAnUnknownProviderNameIsRefusedWithoutEchoingAnythingUnbounded(): void
    {
        $client = $this->client(['providers' => ['default' => $this->signer()]]);

        try {
            $client->sign($this->rendered(), str_repeat('n', 500)."\x00");
            self::fail('the unknown provider was not refused');
        } catch (UsageException $e) {
            self::assertStringContainsString('no signing provider named', $e->getMessage());
            self::assertLessThan(150, strlen($e->getMessage()));
            self::assertStringNotContainsString("\x00", $e->getMessage());
        }
    }

    public function testConfiguredStrictnessSurvivesACallSiteAskingForStrictFalse(): void
    {
        // The ONE place configuration beats a call site: a restriction an
        // operator declared must not be liftable by application code.
        Configuration::configure(['strict' => true]);

        $client = new Client(
            templates: $this->fixtureTemplates(),
            binary: $this->engineBinary(),
            strict: false,
            env: false,
        );

        $this->expectException(UsageException::class);
        $this->expectExceptionMessageMatches('/`generateSource` is disabled/');
        $client->generateSource(template: "version: 0.1.0\nname: x\n");
    }

    public function testANamedProviderIsAcceptedOutsideStrictModeToo(): void
    {
        // Naming providers is good practice everywhere; only the REFUSAL of
        // the alternative belongs to strict.
        $client = $this->client(['providers' => ['default' => $this->signer()]]);

        self::assertTrue($client->sign($this->rendered(), 'default')->success());
    }

    public function testANonStrictClientAcceptsAProviderObject(): void
    {
        self::assertTrue($this->client()->sign($this->rendered(), $this->signer())->success());
    }

    public function testTheLockdownReportsWhetherItIsInForce(): void
    {
        self::assertTrue((new Lockdown(strict: true))->strict());
        self::assertFalse((new Lockdown(strict: false))->strict());
    }
}

File: sdk/java/src/test/java/jp/kengos/shojiku/LockdownTest.java
package jp.kengos.shojiku;

import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.Map;
import org.junit.jupiter.api.Test;

/**
 * The lockdown, one clause at a time.
 *
 * <p>A lockdown tested as a whole reports "something was refused" and stops proving which rule did
 * it. Each clause below is its own test for that reason.
 *
 * <p>Every refusal here is the misuse EXCEPTION rather than a failed result: strict disables an
 * ENTRANCE, so calling it is the program contradicting its own deployment's configuration — not a
 * fact about a document — and a failed result is something {@code if (result.success())} can
 * swallow.
 */
class LockdownTest extends Fixtures {

  private static Map<String, SigningProvider> registry() {
    return Map.of("release", signer());
  }

  @Test
  void aStrictClientRefusesTheTextEntrance() {
    // Clause one: every document a strict client signs came from the configured
    // template root, with its containment rules.
    ShojikuClient client = client().strict(true).providers(registry()).build();

    UsageException error =
        assertThrows(UsageException.class, () -> client.generateSource("version: 0.1.0\n"));

    assertTrue(error.getMessage().contains("generateSource"));
  }

  @Test
  void aStrictClientStillRendersFromItsOwnRoot() {
    assertTrue(
        client().strict(true).providers(registry()).build().generate("receipt", null).success());
  }

  @Test
  void aStrictClientRefusesToSignAnArtifactItDidNotRender() {
    // Clause two. Bytes handed over whole are the caller's, exactly like a
    // caller-supplied template.
    ShojikuClient strict = client().strict(true).providers(registry()).build();
    DocumentArtifact archived = strict.artifact(rendered(client().build()).bytes());

    UsageException error =
        assertThrows(UsageException.class, () -> strict.sign(archived, "release"));

    assertTrue(error.getMessage().contains("loaded"));
  }

  @Test
  void aStrictClientRefusesToSignASourceArtifactToo() {
    // The gap a boolean "was it loaded" would leave open: an artifact from another
    // client's text-first render has engine-laid-out bytes and a caller's template.
    DocumentArtifact fromSource = client().build().generateSource(receiptSource()).unwrap();
    ShojikuClient strict = client().strict(true).providers(registry()).build();

    UsageException error =
        assertThrows(UsageException.class, () -> strict.sign(fromSource, "release"));

    assertTrue(error.getMessage().contains("source"));
  }

  @Test
  void anArtifactAStrictClientWillNotSignIsStillVerifiable() {
    // Clause three, and the deliberate asymmetry: verifying bytes of unknown
    // provenance is the entire point of verify, and a locked-down deployment is
    // precisely the one that must check an archived document it did not produce.
    byte[] archived = signed(client().build()).bytes();
    ShojikuClient strict = client().strict(true).providers(registry()).build();

    assertTrue(
        strict.artifact(archived).verify(java.util.List.of(key("rsa2048.cert.pem"))).success());
  }

  @Test
  void aStrictClientRefusesAProviderObject() {
    // Clause four: signing material must be a provider REGISTERED in configuration,
    // so a key path never appears in request-handling code.
    ShojikuClient strict = client().strict(true).providers(registry()).build();

    UsageException error =
        assertThrows(UsageException.class, () -> strict.sign(rendered(strict), signer()));

    assertTrue(error.getMessage().contains("registered"));
  }

  @Test
  void aStrictClientSignsWithARegisteredName() {
    ShojikuClient strict = client().strict(true).providers(registry()).build();

    assertTrue(strict.sign(rendered(strict), "release").success());
  }

  @Test
  void anUnknownProviderNameIsNamedWithoutEchoingAnythingUnbounded() {
    // Clause five. The name reaches an exception reporter, so it is bounded and
    // stripped exactly as a template name is.
    ShojikuClient strict = client().strict(true).providers(registry()).build();
    String hostile = "z".repeat(500);

    UsageException error =
        assertThrows(UsageException.class, () -> strict.sign(rendered(strict), hostile));

    assertTrue(error.getMessage().contains("no signing provider named"));
    assertFalse(error.getMessage().contains("z".repeat(200)));
  }

  @Test
  void configuredStrictnessSurvivesACallSiteThatAsksForStrictFalse() {
    // Clause six, and the ONE place configuration beats a call site: a restriction
    // an operator declared must not be liftable by application code.
    Configuration.configure(config -> config.withStrict(true).withProviders(registry()));

    ShojikuClient client = client().strict(false).providers(registry()).build();

    assertThrows(UsageException.class, () -> client.generateSource("version: 0.1.0\n"));
  }

  @Test
  void aNameResolvesToARegisteredProviderOutsideStrictModeToo() {
    // Naming providers is good practice everywhere; only the REFUSAL of the
    // alternative is strict's.
    ShojikuClient client = client().providers(registry()).build();

    assertTrue(client.sign(rendered(client), "release").success());
  }

  @Test
  void anUnknownNameIsRefusedOutsideStrictModeAsWell() {
    ShojikuClient client = client().providers(registry()).build();

    assertThrows(UsageException.class, () -> client.sign(rendered(client), "staging"));
  }

  @Test
  void somethingThatIsNotAProviderAtAllIsProgrammerMisuse() {
    ShojikuClient client = client().build();

    UsageException error =
        assertThrows(
            UsageException.class, () -> client.sign(rendered(client), Integer.valueOf(42)));

    assertTrue(error.getMessage().contains("SigningProvider"));
  }

  @Test
  void aNonStrictClientNeedsNoRegistryAtAll() {
    ShojikuClient client = client().build();

    assertTrue(client.sign(rendered(client), signer()).success());
  }
}
